R/bayesian.R

```diff
@@ -0,0 +1,311 @@
+#### Functions for doing Bayesian inference on model runs
+
+#' Get historical land use data
+#'
+#' Fetch the historical land use data and filter to the desired regions, years,
+#' and GCAM commodities.
+#'
+#' All of the runs in a single ensemble should have the same regions, years, and
+#' commodities, so generally it will only be necessary to run this function once
+#' at the beginning of the analysis.  The resulting data can be matched to the
+#' model results for each of the models run in order to calculate Bayesian
+#' likelihoods.
+#'
+#' This function also computes the grouped variance for use in subsequent
+#' calculations.
+#'
+#' @section TODO:
+#'
+#' Calculate obsvar as a detrended variance instead of total variance.
+#'
+#' @param regions Regions to keep in filtered data. (Default is use all regions.)
+#' @param years Years to keep in filtered data. (Default is use all years.)
+#' @param commodities GCAM commodities to keep in filtered data (Default is to
+#' use all commodities.)
+#' @return Data frame containing the filtered data.
+#' @export
+get_historical_land_data <- function(regions = NULL, years = NULL,
+                                     commodities = NULL)
+{
+    ## silence notes
+    GCAM_commodity <- variable <- year <- area <- obsvar <-
+        region <- land.type <- obs <- NULL
+
+    FAO_land_history <- gcamland::FAO_land_history
+
+    filter <- rep(TRUE, nrow(FAO_land_history))
+    if(!is.null(regions))
+        filter <- filter & FAO_land_history$region %in% regions
+    if(!is.null(years))
+        filter <- filter & FAO_land_history$year %in% years
+    if(!is.null(commodities))
+        filter <- filter & FAO_land_history$GCAM_commodity %in% commodities
+
+    FAO_land_history[filter,] %>%
+      dplyr::mutate(variable="Land Area") %>%
+      dplyr::select(region, land.type=GCAM_commodity, variable, year, obs=area) %>%
+      group_by(land.type, variable, region) %>%
+      mutate(obsvar = stats::var(obs)) %>%
+      ungroup
+}
+
+
+#' Load land use results for an already-run scenario
+#'
+#' Fetch land use results and aggregate to region, commodity, and year.
```

Comment:
Is it worth noting that you aren't explicitly grouping by region? It doesn't matter now since we only have one region, but it may be important later.

Reply:
It looks like there is no `region` column go group by in the data currently.  Here's an example:
```
WE27755% head ensemble/land/landAllocation_Perf_0001.csv 
name,land.allocation,year,scenario
UrbanLandAEZ16,0.0049374,1975,Hindcast_Perfect_AgroForest3.05_AgroForestNonPasture3.05_Cropland3.05
UrbanLandAEZ16,0.0049374,1976,Hindcast_Perfect_AgroForest3.05_AgroForestNonPasture3.05_Cropland3.05
```

Reply:
Is it worth noting that in the comment?

Reply:
Ok if I add that in the other branch, just so I don't have to keep triggering tests on this branch?